### DISCUSSION TITLE
Condition when service is first started

### DISCUSSION THREAD

**Discussion by hongkongkiwi** [finit-project/finit#264] [Ideas]
So I'm trying to make syslogd to run first and then everything else run later.

Apart from a plugin, is there anyway in the current configuration to achieve this?

Right now I've setup dependencies, but I found if syslogd goes down or can't start it takes the whole system down. What happens with syslogd is it actually appears in the running state, but then crashes.

So, I was thinking as a solution to this, what about if there was an additional cond that's set for each service. For example `<service/SERVICE/started>` (or something similar) and as long as the service has been started once, this will be set as true. I'm not sure if this is the best name, but basically the idea is once this condition is set it's never unset.

That way, I can have it so that **syslogd** will start first, but then once it's started if it crashes or gets stopped in future for some reason, it won't take down everything else.

**Comment by troglobit** [comment]
I understand what you want to achieve, but conditions weren't really designed for this, as you discovered yourself; crashing or stopping syslogd would bring all dependent services.  I'd suggest using the runlevel trick (below) or the following setup:

For the systems I've been involved in we set up syslogd (the sysklogd project) to always run, i.e. runlevels `[S123456789]` that allows it to start as early as possible. For every script or task that needs to run before that, and does logging, we use the `logger` from sysklogd which supports `-k` -- logging to the kernel ring buffer if syslogd has not yet started up (and created `/dev/log`).  When syslogd starts up it can read everything from the kernel ring buffer and sort it properly to the correct log file(s).  This works fairly well and we've never really had any significant problems with loss of log messages other than when we're debugging the kernel or Finit.

To emulate what I believe systemd does, you can let syslogd start up properly in runlevel 2 and have a script/daemon transition the system to runlevel 3 when it detects syslogd is up.  In runlevel 3 everything else in your system is brought up.

**Reply by hongkongkiwi** [reply]
Does finit start services in runlevel order?

e.g. if I have something like:
```
[12345789] syslogd
[345789] myservice
```

Will finit always start syslogd before myservice if I set runlevel 3 at boot?

**Reply by troglobit** [reply]
If you set runlevel 3 at boot (or in finit.conf), the system will go from S to 3, just like SysV init.

I'm suggesting you set runlevel 2 and then have a task in runlevel 2, that depends on syslogd, which switches the system to runlevel 3 ... or similar.

**Reply by hongkongkiwi** [reply]
Ok, I see. Thanks for the idea, it looks like that might be the best approach.

**Comment by az143** [comment]
i've also come across a requirement for sequencing-only (not-quite-)dependencies on an low-powered embedded box recently, where starting many services in parallel causes thrashing and seriously bad system load. 

like hongkongkiwi i just want to sequence the initial boot steps. as the services in question are independent, adding fake but strict service dependencies is undesirable because restarting one now unnecessarily nukes a whole bunch of them.

personally i've worked around this with user conditions (in 4.5rc5):
normally start first service X, add task that depends on X being ready and then asserts user condition Xready, and have further services Y and Z depend on user condition Xready (and possibly further Yready etc).  tasks aren't rerun, so the asserted Xready doesn't go away and restarting serviceX doesn't affect serviceY or serviceZ.

```
service [2345] serviceX
task [2345] <service/serviceX/ready> name:Xready /sbin/initctl cond set Xready
service [2345] <service/realdependency/ready,usr/Xready> serviceY
```

this feels a bit hackish, but works and doesn't require changes to finit's dependency system.
